### PR TITLE
Add `Directory.sync`

### PR DESCRIPTION
**Pull request by helderco**
Part of https://github.com/dagger/dagger/issues/5065

### CODE REVIEW

File: core/integration/services_test.go [hunk 1]
```diff
@@ -839,6 +839,18 @@ func TestContainerDirectoryServices(t *testing.T) {
 		require.NoError(t, err)
 		require.Equal(t, content, string(fileContent))
 	})
+
+	t.Run("runs services for Container.Directory.Sync", func(t *testing.T) {
```
Comment (vito) [inline core/integration/services_test.go:843]: I think this is only passing because `Container.Directory` "un-lazies" while running the services already.

You could test this by using a Git daemon service instead, using the `git()` API to return a service-dependent Directory, a lot like this test:

https://github.com/dagger/dagger/blob/9ec0d6a9625c85655e8d82f241771470fe6998aa/core/integration/services_test.go#L916-L934

Comment (helderco) [reply core/integration/services_test.go:843]: I wondered why not having `WithServices` passed the test. Thanks for the heads up!

@vito, the `Container.Evaluate` also remarshals from state:

https://github.com/dagger/dagger/blob/9ec0d6a9625c85655e8d82f241771470fe6998aa/core/container.go#L1293-L1301

Is that needed here? I tried making a failing test for not having it but couldn't. Maybe because `Container` just has more moving pieces? Still good to have here anyway for future proofing? Can you explain why it's needed?

Comment (vito) [reply core/integration/services_test.go]: My best guess is that it's to deal with the `scratch` case where `FS` will be `nil`, but it might not be necessary, and we could always detect `nil` and do a no-op instead. Could be worth a refactor if you want to also add a test for the scratch cases for Directory and Container.

Comment (helderco) [reply core/integration/services_test.go]: Yes, but same as before, I started with that but couldn't make a failing test. Any idea how?

You can see I added the "empty" test with: `client.Directory().Sync(ctx)`. `Query.Directory` already marshals `llb.Scratch()`:

https://github.com/dagger/dagger/blob/9ec0d6a9625c85655e8d82f241771470fe6998aa/core/schema/directory.go#L81

https://github.com/dagger/dagger/blob/9ec0d6a9625c85655e8d82f241771470fe6998aa/core/directory.go#L54-L61

Comment (vito) [reply core/integration/services_test.go]: This comment seems related:

https://github.com/dagger/dagger/blob/e72d705eda04f30e5a89d0b93a3469ec12859031/core/util.go#L62-L68

The tl;dr is marshaling `llb.Scratch()` results in a wonky `*pb.Definition` that causes errors in Buildkit, so we try to work around that by assigning `nil` instead of a marshalled `llb.Scratch()` in places like `Container.FS` and `Directory.LLB`. It's kind of a pain.

`Container.FSState` calls this under the hood, so maybe this is attempting to be durable to cases where a `llb.Scratch()` ends up being marshaled to `Container.FS` instead of setting `Container.FS` to `nil`.

`Query.Directory` could probably use `core.NewScratchDirectory` instead, which didn't exist at the time. It's possible that just nothing fails here either way, but calling `NewScratchDirectory` feels a bit cleaner since it sets `LLB` to `nil` instead of marshaling `llb.Scratch()`.

Hope this helps, but it might just raise more questions. :sweat_smile:

Comment (helderco) [reply core/integration/services_test.go]: Thanks for the clarification but the question remains: how do I make a failing test **for not doing** the marshal (see https://github.com/dagger/dagger/pull/5414#discussion_r1255737022) and a no-op on nil?

Comment (helderco) [reply core/integration/services_test.go]: > It's possible that just nothing fails here either way

Ah, I missed this. Yes, that's what happens. I assume because this doesn't error:

```go
		gw.Solve(ctx, bkgw.SolveRequest{
			Evaluate:   true,
			Definition: nil,
		})
```

Comment (helderco) [reply core/integration/services_test.go]: @vito I switched to `NewScratchDirectory` and added the no-op. Now going back to the marshal in `Container.Evaluate`, is there a need to do it here?

**Review by vito — APPROVED** [review]
lgtm!

File: core/integration/services_test.go (hunk 1 above)
Comment (vito) [reply core/integration/services_test.go]: > ~Now going back to the marshal in `Container.Evaluate`, is there a need to do it here? I suppose not, from your explanation, but just to confirm.~

Might be OK to remove it, but I'm ok with letting sleeping dogs lie. :grin:
